### ISSUE TITLE
please add sentence and paragraph text objects (overide)

### ISSUE THREAD

**Issue by yashamon** [lervag/vimtex#3041]
The built in sentence and paragraph don't work well for latex for example if I have a block of the form

```tex
\begin{equation}
blah.
blah.
\end{equation} 
Here is my sentence.
```

Then doing `vis` on the bottom sentence also selects `end{equation}`.

**Comment by kiryph** [comment]
I think your issue has already been raised. See old issues from 2016:

- https://github.com/lervag/vimtex/issues/320
- https://github.com/lervag/vimtex/issues/437

If I read the old issues correctly, a text object for sentences has been rejected.

Sentence detection is more complicated than one obvious case might suggest. 

Without LaTeX code see: https://en.wikipedia.org/wiki/Sentence_boundary_disambiguation

But with LaTeX code you would need to distinguish commands and inline environments which can occur within a sentence (e.g. [tikz](https://tex.stackexchange.com/questions/313927/tikz-picture-inline)) and those which separate sentences.

### Sentence tokenization with machine learning

You might consider machine-learning approaches. I have run sentence tokenization on plain text without LaTeX code with existing training data here:

https://github.com/lervag/vimtex/issues/544#issuecomment-299444814

But I think you need training data for sentence tokenization with interlaced latex code using many different packages which I do not know exists.

### Treesitter syntax tree

The treesitter syntax tree in neovim does not do sentence tokenization either but would separate equations.

Maybe you only need a possibility to select blocks of text without latex commands or environments.

Look at following screenshot of a syntax tree:
![Screenshot 2024-12-02 at 17 01 24](https://github.com/user-attachments/assets/34973f98-8062-4330-b1b0-ccaa1514f72d)

You can consider using incremental selection from https://github.com/nvim-treesitter/nvim-treesitter?tab=readme-ov-file#incremental-selection or a customization of 

https://github.com/nvim-treesitter/nvim-treesitter-textobjects?tab=readme-ov-file#overriding-or-extending-textobjects

e.g. for a text block.

**Comment by lervag** [comment]
> The built in sentence and paragraph don't work well for latex for example if I have a block of the form …

Sorry for the very late reply!

I think @kiryph is quite right in his response:

> I think your issue has already been raised. See old issues from 2016: …
> 
> If I read the old issues correctly, a text object for sentences has been rejected.
> 
> Sentence detection is more complicated than one obvious case might suggest.

I believe this is the main point here. It is actually surprisingly hard to parse LaTeX, and thus it is very hard to specify correct sentence boundaries. I've personally settled on the pragmatic choice of just accepting this and "moving on".

Nonetheless, if you or anyone else would be interested in attempting to implementing this and contributing, then I would be glad to review and participate in the discussion. But I won't personally work on this myself.

I'll close the issue as "won't do", but feel free to reopen if you want to take a stab at it yourself.

**Comment by yashamon** [comment]
I am using the following work around: the snippets for environments contain commented periods. Like this: 
\end{some environment} %.
